File: frontend/types/data/tweet.ts
import {User} from "./user";

export enum TweetTypes {
    Tweet = 0,
    ReTweet,
    QouteTweet,
    Replay,
}

export interface Tweet {
    tweetId: string;
    title: string;
    content: string;
    createdAt: Date;
    hashtags: string[];
    images: string[];
    likesCount: number;
    reTweetType: TweetTypes;
    sender: User;
    video: string;
}

// {
//     "tweetId": "3b328311-50d6-4a63-8297-b90e2b12cf84",
//     "reTweetType": 0,
//     "title": "",
//     "content": "",
//     "images": [],
//     "video": "",
//     "likesCount": 0,
//     "createdAt": "0001-01-01T00:00:00",
//     "hashtags": [],
//     "sender": {
//         "userId": "00000000-0000-0000-0000-000000000000",
//         "fullName": "",
//         "userName": "",
//         "email": "",
//         "birthDay": null,
//         "bio": "",
//         "profileImage": "",
//         "backgroundImage": "",
//         "status": 0,
//         "joinedAt": null,
//         "statusText": "Happy",
//         "userType": 5,
//         "followingsCount": 0,
//         "followersCount": 0,
//         "interestedCategories": []
//     }
// }

File: backend/API.DataLayer/Models/Tweet.cs
namespace API.DataLayer.Models;

public class Tweet
{
    [Key]
    public Guid TweetId { get; set; } = Guid.NewGuid();
    public TweetTypes ReTweetType { get; set; } = TweetTypes.Tweet;

    public bool IsDeleted { get; set; } = false;
    public string? Title { get; set; } = string.Empty;
    [Required]
    public string Content { get; set; } = string.Empty;

    public List<string>? Images { get; set; } = new List<string>();
    public string? Video { get; set; } = string.Empty;

    public Guid? BaseTweetId { get; set; }
    public Tweet? BaseTweet { get; set; }
    public DateTime CreatedAt { get; set; }
    public List<string>? Hashtags { get; set; } = new List<string>();

    [Required]
    public Guid? SenderId { get; set; }
    public User? Sender { get; set; } = new();
    public ICollection<Tweet> Replies { get; set; } = new List<Tweet>() { };
    public ICollection<User> Likers { get; set; } = new List<User>() { };

    public bool IsNull() => string.IsNullOrEmpty(Content);

}

public enum TweetTypes
{
    Tweet=0,
    ReTweet,
    QouteTweet,
    Reply
}